File: MAOOAM/maooam/maooam.py
"""
    Principal module
    ======================

    Python implementation of the Modular Arbitrary-Order Ocean-Atmosphere Model MAOOAM

    .. note :: The python code is available here : \
    `maooam.py <../_modules/maooam.html>`_ .


    :Example:

    >>> from maooam import *

    Global variable
    -------------------

    * **ic.X0** : initial conditions
    * **X** : live step vector
    * **t** : time
    * **t_trans**, **t_run** : respectively transient and running time
    * **dt** : step time
    * **tw** : step time for writing on evol_field.dat

    Dependencies
    -------------------

    >>> import numpy as np
    >>> import params_maooam
    >>> from params_maooam import ndim,tw,t_run,t_trans,dt
    >>> import aotensor
    >>> import time
    >>> import ic_def
    >>> import ic
    >>> import sys
"""

import numpy as np
import params_maooam
from params_maooam import ndim, tw, t_run, t_trans, dt
import integrator
import time
import ic_def
import ic
import sys

def print_progress(p):
    sys.stdout.write('Progress {:.2%} \r'.format(p))
    sys.stdout.flush()

class bcolors:
    """to color the instructions in the console"""
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'


print (bcolors.OKBLUE + "Model MAOOAM v1.3" + bcolors.ENDC)
print (bcolors.OKBLUE + "Initialization ..." + bcolors.ENDC)

ic_def.load_IC()

X = ic.X0
print (bcolors.OKBLUE + "Starting the transient time evolution ..." + bcolors.ENDC)
t = 0.
T = time.clock()
t_up = dt/t_trans*100
while t < t_trans:
    X = integrator.step(X, t, dt)
    t += dt
    if t/t_trans*100 % 0.1 < t_up:
        print_progress(t/t_trans)

print (bcolors.OKBLUE + "Starting the time evolution ..." + bcolors.ENDC)
fichier = open("evol_field.dat", "w")
t = 0.
t_up = dt/t_run*100

while t < t_run:
    X = integrator.step(X, t, dt)
    t += dt
    if t % (tw) < dt:
        fichier.write(str(t)+" ")
        for i in range(0, ndim):
            fichier.write(str(X[i])+" ")
        fichier.write("\n")
    if t/t_run*100 % 0.1 < t_up:
        print_progress(t/t_run)
fichier.close()
print (bcolors.OKBLUE + "Evolution finished " + bcolors.ENDC)

print (bcolors.OKBLUE + "Time clock :" + bcolors.ENDC)
print (time.clock()-T)

File: MAOOAM/maooam_fortran/maooam.f90

!  maooam.f90
!
!> Fortran 90 implementation of the modular arbitrary-order ocean-atmosphere 
!> model MAOOAM.
!
!> @copyright                                                               
!> 2015 Lesley De Cruz & Jonathan Demaeyer.
!> See LICENSE.txt for license information.                                  
!
!---------------------------------------------------------------------------!

PROGRAM maooam 
  USE params, only: ndim, dt, tw, t_trans, t_run, writeout
  USE aotensor_def, only: init_aotensor
  USE IC_def, only: load_IC, IC
  USE integrator, only: init_integrator,step
  USE stat
  IMPLICIT NONE

  REAL(KIND=8), DIMENSION(:), ALLOCATABLE :: X       !< State variable in the model
  REAL(KIND=8), DIMENSION(:), ALLOCATABLE :: Xnew    !< Updated state variable
  REAL(KIND=8) :: t=0.D0                             !< Time variable
  REAL(KIND=8) :: t_up

  PRINT*, 'Model MAOOAM v1.3'
  PRINT*, 'Loading information...'

  CALL init_aotensor    ! Compute the tensor

  CALL load_IC          ! Load the initial condition

  CALL init_integrator  ! Initialize the integrator

  t_up=dt/t_trans*100.D0

  IF (writeout) OPEN(10,file='evol_field.dat')

  ALLOCATE(X(0:ndim),Xnew(0:ndim))

  X=IC

  PRINT*, 'Starting the transient time evolution...'

  DO WHILE (t<t_trans)
     CALL step(X,t,dt,Xnew)
     X=Xnew
     IF (mod(t/t_trans*100.D0,0.1)<t_up) WRITE(*,'(" Progress ",F6.1," %",A,$)') t/t_trans*100.D0,char(13)
  END DO

  PRINT*, 'Starting the time evolution...'

  CALL init_stat
  
  t=0.D0
  t_up=dt/t_run*100.D0

  IF (writeout) WRITE(10,*) t,X(1:ndim)

  DO WHILE (t<t_run)
     CALL step(X,t,dt,Xnew)
     X=Xnew
     IF (mod(t,tw)<dt) THEN
        IF (writeout) WRITE(10,*) t,X(1:ndim)
        CALL acc(X)
     END IF
     IF (mod(t/t_run*100.D0,0.1)<t_up) WRITE(*,'(" Progress ",F6.1," %",A,$)') t/t_run*100.D0,char(13)
  END DO

  PRINT*, 'Evolution finished.'

  IF (writeout) CLOSE(10)

  IF (writeout) OPEN(10,file='mean_field.dat')

  X=mean()
  IF (writeout) WRITE(10,*) X(1:ndim)
  IF (writeout) CLOSE(10)

END PROGRAM maooam 
